src/framework.js

```diff
@@ -0,0 +1,143 @@
+/**
+ * File: framework.js 
+ */
+var _ = require('underscore');
+var util = require('util');
+
+var componentFactory = require('./noflo-component-factory');
+var stateFactory = require('./create-state');
+var promiseOutput = require('../src/promise-output');
+var vniManager = require('./vni-manager');
+
+/**
+ * This module implements the core of the RDF Pipeline Framework.  
+ */
+module.exports = {
+
+    componentFactory: function( nodeDef, fRunUpdater ) { 
+        
+      return componentFactory(
+          _.defaults( { // Add an ondata function to be invoked every time we receive new data
+                        inPorts: _.mapObject(nodeDef.inPorts, function( port, portName ) {  
+                            return _.extend({
+                                ondata: _.partial( ondata, fRunUpdater ) 
+                            }, port);
+                        }),
+                        outPorts: _.extend( {},
+                                            promiseOutput.outPorts,
+                                            nodeDef.outPorts )
+                      },
+                      nodeDef
+          ), 
+          vniManager 
+      );
+    }
+
+}; // module.exports
+
+
+// @this port context
+var ondata = function( fRunUpdater, payload, socketIndex ) {
+
+     var portName = this.name;
+
+     var vnid = payload.vnid || '';
+     var vni = this.nodeInstance.vni(vnid);
+
+     // Check if we have a multi/addressable port and call the correct input-state API
+     if ( _.isUndefined( socketIndex ) ) { 
```

Comment:
No need to do this the input-states.js will handle an undefined socketId, just use full form.